$
\text{Margin distance} = \frac{\vec{w}^{T} \vec{x} + b + \vec{1} - (\vec{w}^{T} \vec{x} + b - \vec{1})}{||\vec{w}||} = \frac{2}{||\vec{w}||}
$

$
\text{max} \frac{2}{||\vec{w}||}\Rightarrow \text{min} ||\vec{w}|| \Rightarrow \text{min} \frac{||\vec{w}||^2}{2}\\
\Rightarrow \text{argmin}_{\vec{w}} \frac{1}{2} ||\vec{w}||^2\ \text{such that}\
1\le y_i (\vec{w}^{T} \vec{x_i} + b)\ \forall i\in \{1,...,N\}
$

In [1]:
import numpy as np

In [9]:
x1 = np.array([1,2,3])
x2 = np.array([4,1,2])
x3 = np.array([-1,2,-1])

y1 = 1
y2 = 1
y3 = -1

# A
print('A.')
wa = np.array([.3, 0, .4])
b = -.4

print('For y1:', y1*(np.dot(wa, x1)+b))
print('For y2:', y2*(np.dot(wa, x2)+b))
print('For y3:', y3*(np.dot(wa, x3)+b))

# B
print('B.')
wb = np.array([.2, 0, .4])
b = -.4

print('For y1:', y1*(np.dot(wb, x1)+b))
print('For y2:', y2*(np.dot(wb, x2)+b))
print('For y3:', y3*(np.dot(wb, x3)+b))

# C
print('C.')
wc = np.array([.1, 0, .4])
b = -.4

print('For y1:', y1*(np.dot(wc, x1)+b))
print('For y2:', y2*(np.dot(wc, x2)+b))
print('For y3:', y3*(np.dot(wc, x3)+b))

# D
print('D.')
wd = np.array([.4, 0, .2])
b = -.4

print('For y1:', y1*(np.dot(wd, x1)+b))
print('For y2:', y2*(np.dot(wd, x2)+b))
print('For y3:', y3*(np.dot(wd, x3)+b))

# compare A and B
print('\nCompare the objective for A and B')
print('A:', (np.linalg.norm(wa)**2)/2)
print('B:', (np.linalg.norm(wb)**2)/2)

A.
For y1: 1.1
For y2: 1.6
For y3: 1.1
B.
For y1: 1.0
For y2: 1.2000000000000002
For y3: 1.0
C.
For y1: 0.9000000000000002
For y2: 0.8000000000000002
For y3: 0.9
D.
For y1: 0.6
For y2: 1.6
For y3: 1.0

Compare the objective for A and B
A: 0.125
B: 0.10000000000000002


$
\frac{dg(z)}{dz} = \begin{cases}
1 & z > 0\\
0 & z < 0\\
\text{undefined} & z = 0
\end{cases}
$

$
\because z_i = -1\\
\therefore h_1'(z_i) = 0,\ \text{according to problem 4.(c) in SVM}\\
\therefore \text{any weight derivative including $\frac{dh_1(z_i)}{z_i}$ should be $0$}\\
\because \text{based on the network, $\frac{\partial C}{\partial w_2}$ and $\frac{\partial C}{\partial w_3}$ includes $\frac{dh_1(z_i)}{z_i}$}\\
\therefore \text{$\frac{\partial C}{\partial w_2}$ and $\frac{\partial C}{\partial w_3}$ are gauranteed to be $0$ while the value of $\frac{\partial C}{\partial w_1}$ is not sure}
$

In [4]:
def sigmoid(x):
    return 1/(1 + np.exp(-x))

x = np.array([1,4,5])
w_h = np.array([[.1,.2],
                [.3,.4],
                [.5,.6]])
w_o = np.array([[.7,.9],
                [.8,.1]])
b_1 = np.array([.5,.5])
b_2 = np.array([.5,.5])

zh = np.dot(x, w_h) + b_1
zo = np.dot(zh, w_o) + b_2

print('h1\n', sigmoid(zh[0]))
print('o1\n', sigmoid(zo[0]))

h1
 0.9866130821723351
o1
 0.9995694429186754


$
\frac{\partial E}{\partial w_7} = \frac{\partial E}{\partial o_1}\frac{\partial o_1}{\partial zo_1}\frac{\partial zo_1}{\partial w_7}\\
\ \\
\frac{\partial E}{\partial b_2} = \frac{\partial E}{\partial o_1}\frac{\partial o_1}{\partial zo_1}\frac{\partial zo_1}{\partial b_2} +
\frac{\partial E}{\partial o_2}\frac{\partial o_2}{\partial zo_2}\frac{\partial zo_2}{\partial b_2}\\
\ \\
\frac{\partial E}{\partial w_1} = \frac{\partial E}{\partial o_1}\frac{\partial o_1}{\partial zo_1}\frac{\partial zo_1}{\partial h_1} 
\frac{\partial h_1}{\partial zh_1}\frac{\partial zh_1}{\partial w_1}
+
\frac{\partial E}{\partial o_2}\frac{\partial o_2}{\partial zo_2}\frac{\partial zo_2}{\partial h_1} 
\frac{\partial h_1}{\partial zh_1}\frac{\partial zh_1}{\partial w_1}
$

In [14]:
.1 - .01*0.00127152

0.0999872848

$
\because \frac{\partial E}{\partial o_1} = o_1 - t_1 = 0.89956944,\
\frac{\partial E}{\partial o_2} = o_2 - t_2 = 0.75\\
\because \frac{\partial o_1}{\partial zo_1} = o_1(1-o_1) = 0.00043037,\
\frac{\partial o_2}{\partial zo_2} = o_2(1-o_2) = 0.16\\
\because \frac{\partial zo_1}{\partial w_7} = h_1 = 0.98661308,\
\frac{\partial zo_1}{\partial b_2} = 1,\
\frac{\partial zo_2}{\partial b_2} = 1\\
\because \frac{\partial zo_1}{\partial h_1} = w_7 = 0.7,\
\frac{\partial zo_2}{\partial h_1} = w_8 = 0.8\\
\because \frac{\partial h_1}{\partial zh_1} = h_1(1-h_1) = 0.013207710,\
\frac{\partial zh_1}{\partial w_1} = x_1 = 1\\
\therefore \frac{\partial E}{\partial w_7} = 0.00038196,\
\frac{\partial E}{\partial b_2} = 0.12038715,\
\frac{\partial E}{\partial w_1} = 0.00127152\\
\therefore w_7^+ = w_7 - \alpha \times \frac{\partial E}{\partial w_7} = 0.69999618\\
\therefore b_2^+ = b_2 - \alpha \times \frac{\partial E}{\partial b_2} = 0.49879613\\
\therefore w_1^+ = w_1 - \alpha \times \frac{\partial E}{\partial w_1} = 0.09998728\\
$

In [18]:
1 - (-1)*np.dot([2,2]-np.linalg.norm([2,2]), [-1,-1])-1

1.6568542494923806

$
\Rightarrow C \vec{w} + \sum_{i=1}^{M} -y_i \vec{x_i},\ \text{where}\ x_i\in \{x_1,\ x_2,\ ...,\ x_M\}\ \text{satisfy}\ 1-y_i(\vec{w}^T \vec{x}_i +b) > 0
$

$
\because 1-y_1(\vec{w}^T \vec{x_1} +b) = -2 < 0\\
\therefore \nabla \vec{w} = \begin{bmatrix}
2\\ 2
\end{bmatrix}\\
\therefore \vec{w}_1 = \vec{w} - 1\times \nabla \vec{w} = \begin{bmatrix}
0\\ 0
\end{bmatrix}\\
\because 1-y_2(\vec{w}^T \vec{x_2} +b) = 0 \\
\therefore \nabla \vec{w} = \begin{bmatrix}
0\\ 0
\end{bmatrix} +
\vec{x_2}\\
\therefore \vec{w}_2 = \vec{w_1} - 1\times \nabla \vec{w} = \begin{bmatrix}
1\\ 1
\end{bmatrix}
$